File: app/src/main/java/com/yhjoo/dochef/activities/FollowListActivity.java
package com.yhjoo.dochef.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.LinearLayoutManager;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.google.gson.JsonObject;
import com.yhjoo.dochef.App;
import com.yhjoo.dochef.R;
import com.yhjoo.dochef.adapter.FollowListAdapter;
import com.yhjoo.dochef.databinding.AFollowlistBinding;
import com.yhjoo.dochef.interfaces.RxRetrofitServices;
import com.yhjoo.dochef.model.UserBrief;
import com.yhjoo.dochef.model.UserDetail;
import com.yhjoo.dochef.utils.DataGenerator;
import com.yhjoo.dochef.utils.RxRetrofitBuilder;
import com.yhjoo.dochef.utils.Utils;

import java.util.ArrayList;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.functions.Function;
import retrofit2.Response;

public class FollowListActivity extends BaseActivity {
    public enum MODE {FOLLOWER, FOLLOWING}

    AFollowlistBinding binding;
    RxRetrofitServices.UserService rxUserService;
    FollowListAdapter followListAdapter;

    MODE current_mode = MODE.FOLLOWER;
    UserDetail userDetailInfo;
    ArrayList<UserBrief> userList;

    String active_userid;
    String target_id;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = AFollowlistBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.followlistToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        rxUserService = RxRetrofitBuilder.create(this, RxRetrofitServices.UserService.class);

        active_userid = Utils.getUserBrief(this).getUserID();
        target_id = getIntent().getStringExtra("userID");
        current_mode = (MODE) getIntent().getSerializableExtra("MODE");

        binding.followlistToolbar.setTitle(current_mode == MODE.FOLLOWER ? "팔로워" : "팔로잉");
        followListAdapter = new FollowListAdapter(active_userid);
        followListAdapter.setEmptyView(R.layout.rv_loading, (ViewGroup) binding.followlistRecycler.getParent());
        followListAdapter.setOnItemClickListener((adapter, view, position) -> {
                    Utils.log(((UserBrief) adapter.getData().get(position)).getUserID());
                    Intent intent = new Intent(FollowListActivity.this, HomeActivity.class)
                            .putExtra("userID", ((UserBrief) adapter.getData().get(position)).getUserID());
                    startActivity(intent);
                }
        );
        followListAdapter.setOnItemChildClickListener(this::onListItemClick);
        binding.followlistRecycler.setLayoutManager(new LinearLayoutManager(this));
        binding.followlistRecycler.setAdapter(followListAdapter);

        if (App.isServerAlive()){
            Single<Response<ArrayList<UserBrief>>> modeSingle;
            if (current_mode == MODE.FOLLOWER)
                modeSingle = rxUserService.getFollowers(target_id)
                        .observeOn(AndroidSchedulers.mainThread());
            else
                modeSingle = rxUserService.getFollowings(target_id)
                        .observeOn(AndroidSchedulers.mainThread());

            compositeDisposable.add(
                    rxUserService.getUserDetail(active_userid)
                            .flatMap((Function<Response<UserDetail>, Single<Response<ArrayList<UserBrief>>>>)
                                    response -> {
                                        userDetailInfo = response.body();
                                        return modeSingle;
                                    }
                            )
                            .subscribe(response->{
                                userList = response.body();
                                setListData();
                            }, RxRetrofitBuilder.defaultConsumer())
            );
        }
        else {
            userDetailInfo = DataGenerator.make(getResources(), getResources().getInteger(R.integer.DATA_TYPE_USER_DETAIL));
            userList = DataGenerator.make(getResources(), getResources().getInteger(R.integer.DATA_TYPE_USER_BRIEF));

            setListData();
        }
    }

    void onListItemClick(BaseQuickAdapter adapter, View view, int position) {
        String target = ((UserBrief) adapter.getData().get(position)).getUserID();

        Single<Response<JsonObject>> subORunsub = view.getId() == R.id.user_followcancel_btn ?
                rxUserService.subscribeUser(active_userid, target):
                rxUserService.unsubscribeUser(active_userid, target);

        Single<Response<ArrayList<UserBrief>>> modeSingle = current_mode == MODE.FOLLOWER ?
                rxUserService.getFollowers(target_id)
                        .observeOn(AndroidSchedulers.mainThread()):
                rxUserService.getFollowings(target_id)
                        .observeOn(AndroidSchedulers.mainThread());

        compositeDisposable.add(
                subORunsub
                        .flatMap((Function<Response<JsonObject>, Single<Response<UserDetail>>>)
                                response -> rxUserService.getUserDetail(active_userid))
                        .flatMap((Function<Response<UserDetail>, Single<Response<ArrayList<UserBrief>>>>)
                                response -> {
                                    userDetailInfo = response.body();
                                    return modeSingle;
                                }
                        )
                        .subscribe(response->{
                            userList = response.body();
                            setListData();
                        }, RxRetrofitBuilder.defaultConsumer())
        );
    }

    void setListData() {
        followListAdapter.setActiveUserFollow(userDetailInfo.getFollow());

        followListAdapter.setNewData(userList);
        followListAdapter.setEmptyView(R.layout.rv_empty, (ViewGroup) binding.followlistRecycler.getParent());
    }
}

File: app/src/main/java/com/yhjoo/dochef/ui/activities/FollowListActivity.kt
package com.yhjoo.dochef.ui.activities

import android.content.Intent
import android.os.Bundle
import android.view.*
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import com.yhjoo.dochef.R
import com.yhjoo.dochef.adapter.FollowListAdapter
import com.yhjoo.dochef.databinding.FollowlistActivityBinding
import com.yhjoo.dochef.model.UserBrief
import com.yhjoo.dochef.repository.FollowListRepository
import com.yhjoo.dochef.utilities.*
import com.yhjoo.dochef.viewmodel.FollowListViewModel
import com.yhjoo.dochef.viewmodel.FollowListViewModelFactory
import java.util.*

class FollowListActivity : BaseActivity() {
    object UIMODE {
        const val FOLLOWER = 0
        const val FOLLOWING = 1
    }

    private val binding: FollowlistActivityBinding by lazy {
        DataBindingUtil.setContentView(this, R.layout.followlist_activity)
    }

    private lateinit var followlistViewModel: FollowListViewModel

    private lateinit var followListAdapter: FollowListAdapter
    private lateinit var activeUserId: String
    private lateinit var currentUserId: String

    private var currentMode = UIMODE.FOLLOWER

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        setSupportActionBar(binding.followlistToolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        currentMode = intent.getIntExtra("MODE", UIMODE.FOLLOWER)
        activeUserId = Utils.getUserBrief(this).userID
        currentUserId = intent.getStringExtra("userID").toString()

        val factory = FollowListViewModelFactory(
            FollowListRepository(
                applicationContext,
                currentMode,
                activeUserId,
                currentUserId
            )
        )
        followlistViewModel = factory.create(FollowListViewModel::class.java).apply {
            activeUserDetail.observe(this@FollowListActivity, {
                followListAdapter.activeUserFollowList = it.follow
                followListAdapter.notifyDataSetChanged()
            })
            allFollowLists.observe(this@FollowListActivity, {
                followListAdapter.submitList(it) {
                    binding.followlistRecycler.scrollToPosition(0)
                }
            })
        }

        binding.apply {
            lifecycleOwner = this@FollowListActivity

            followlistToolbar.title = if (currentMode == UIMODE.FOLLOWER) "팔로워" else "팔로잉"

            followListAdapter = FollowListAdapter(
                activeUserId,
                { item -> followlistViewModel.subscribeUser(item.userID) },
                { item -> followlistViewModel.unsubscribeUser(item.userID) },
                { item -> itemClicked(item) },
            )

            followlistRecycler.apply {
                layoutManager = LinearLayoutManager(this@FollowListActivity)
                adapter = followListAdapter
            }
        }
    }

    private fun itemClicked(userBrief: UserBrief): Unit {
        val intent = Intent(this@FollowListActivity, HomeActivity::class.java)
            .putExtra("userID", userBrief.userID)
        startActivity(intent)
    }
}